### PR TITLE
add list() and exists() to filestore

### PR DESCRIPTION
**Pull request by Alexey-N-Chernyshov**
### Description of the Change
Add list() and exists() functions to Filestore.
Change pointers to gsl::span in interfaces.

### Benefits
Added functionality.

### Possible Drawbacks 
No.

### CODE REVIEW

File: core/storage/filestore/impl/filesystem/filesystem_file.cpp [hunk 1]
```diff
@@ -37,40 +37,42 @@ fc::outcome::result<void> FileSystemFile::open() noexcept {
```
Comment (masterjedy) [inline core/storage/filestore/impl/filesystem/filesystem_file.cpp]: in previous projects we used to follow this scheme:
root:
| file.hpp // interface
| impl/file_impl.hpp // implementation
| impl/file_impl.cpp //implementation

and what concerns naming:
class File {virtual some_fun() = 0;}; // interface
class FileImpl: public File { some_fun() override {...} }; // implementation

File: core/storage/filestore/impl/filesystem/filesystem_file.cpp [hunk 2]
```diff
@@ -37,40 +37,42 @@ fc::outcome::result<void> FileSystemFile::open() noexcept {
 
 fc::outcome::result<void> FileSystemFile::close() noexcept {
```
Comment (masterjedy) [inline core/storage/filestore/impl/filesystem/filesystem_file.cpp]: And by the way, storing folded filesystem directory is redundant.

File: test/core/storage/filestore/filesystem/filesystem_file_test.cpp [hunk 3]
```diff
@@ -118,35 +114,44 @@ TEST_F(FileSystemFileTest, CloseFile) {
   ASSERT_EQ(0, res_size.value());
 }
 
+/**
+ * @given file exists
```
Comment (masterjedy) [inline test/core/storage/filestore/filesystem/filesystem_file_test.cpp]:
```suggestion
 * @given an existing file
```

File: test/core/storage/filestore/filesystem/filesystem_file_test.cpp [hunk 4]
```diff
@@ -190,35 +192,44 @@ TEST_F(FileSystemFileTest, WriteAtPos) {
...
-  ASSERT_EQ(more_data_size, write_res.value());
-
-  char expected[] = "hello beauty C++ world";
-  size_t expected_size = strlen(expected);
+  std::vector<char> more_data{'C', '+', '+', ' ', 'w', 'o', 'r', 'l', 'd'};
+
+  EXPECT_OUTCOME_TRUE(write_res2, empty_file->write(start, more_data));
+  ASSERT_EQ(more_data.size(), write_res2);
+
+  std::vector<char> expected{'h',
```
Comment (masterjedy) [inline test/core/storage/filestore/filesystem/filesystem_file_test.cpp]: pls fix formatting

File: core/storage/filestore/impl/filesystem/filesystem_file.cpp (hunk 1 above)
Comment (Alexey-N-Chernyshov) [reply core/storage/filestore/impl/filesystem/filesystem_file.cpp]: Well, it is a specific implementation on boost::filesystem, there are more possible implementations like C files, iostream and etc. I want to reflect the specific implementation in naming.

**Pull request merged**